### ISSUE TITLE
[Filter Panel]: Consolidated accessibility issues

### ISSUE THREAD

**Issue by andy-blum** [carbon-design-system/carbon-for-ibm-dotcom#10589]
## Description

Following a meeting to review the accessibility of the [Filter Panel](https://carbon-design-system.github.io/carbon-for-ibm-dotcom/canary/web-components/iframe.html?args=&id=components-filter-panel--default&viewMode=story), the issues below have been identified.

---

## ❓Sets of checkboxes need to be reachable using arrow keys

<img width="385" alt="Screenshot 2023-06-22 at 12 00 29 PM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/25532785/48001712-fb2c-4149-913f-947d545d3ce9">

According to Ram, the checkboxes in the filter panel need to be navigable using the arrow keys to move through the set of all options. The [ARIA Authoring Practices Guide](https://www.w3.org/WAI/ARIA/apg/patterns/checkbox/examples/checkbox/), however, does not seem to match this assessment:

| Key   | Function                                               |
|-------|--------------------------------------------------------|
| Tab   | Moves keyboard focus to the   checkbox.                |
| Space | Toggles checkbox between checked and unchecked states. |

It seems @mbgower also recommends against arrow-key navigation in #10132 per Olivia's comment.

---

## ❌ Non-checkbox filter buttons are announced as "frames" by JAWS

I'm unable to test with JAWS and VoiceOver isn't announcing this quite the same, so I'm sort of guessing here, but I _think_ this might be related to the use of `aria-controls`.

On the top-level accordion items (Product Types, Technologies, etc), we have this markup:

<img width="1168" alt="Screenshot 2023-06-22 at 12 11 45 PM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/25532785/4a5c1886-00c0-4979-a42a-e42acde47a30">

On the second-level non-checkbox items (like those under Technologies), we have this markup:

<img width="1213" alt="Screenshot 2023-06-22 at 12 11 22 PM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/25532785/815b40ef-0208-435e-b45f-0f8004c9a9dd">

Note that both use the `aria-controls` attribute, but the first one also uses `aria-expanded`. I'm wondering if the use of `aria-controls` _without_ `aria-expanded` is a pattern used to imply something else. On the other hand, Technologies => Artificial Intelligence _does_ have child items and the `aria-expanded` attribute. Perhaps the problem is related to using `<button>` vs `<div role="button">`?

Either way, every single second-level item makes use of the `aria-controls` attribute, even when there are no subitems. We should probably fix that.

---

## ❓Nested filter options are all announced when they are expanded

Again, this seems to be a JAWS-specific bug that I can't duplicate in VO. During the meeting, Ram showed that when a second-level item with children was expanded, all of the list items were read out immediately.

---

## ❌ Mobile version's popup doesn't transfer focus correctly

On iOS voiceover, the filter panel is placed in a modal overlay, but doesn't transfer focus into the modal or trap focus within the modal. This may or may not be an issue specific to the filter panel.

![IMG_2147](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/25532785/fdfc1aa7-990c-4c12-99a2-80eed39c96cd)

![IMG_2148](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/25532785/f46398b9-3216-4c32-bf09-ed13d0f98a7f)

---

## ❌ 🎨 The placement/design of the "clear" button needs to be re-examined.

<img width="388" alt="Screenshot 2023-06-22 at 12 22 50 PM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/25532785/9db44d0d-e020-4175-8028-2a89d68e0ec4">

When items are selected and filters are applied, a new button pops into the DOM allowing users to clear their selections. The location of this button, however, places it in a location that's not intuitive to assistive technology users, as they'd have to search in reverse to find it. We have two primary options to fix this:

1. Move the clear-selections button to the bottom of the filter panel. This would make it easier for AT users to discover the button after making their selections
2. Keep the clear-selections button in the DOM but disable it when no filters are applied. In this way, AT users can scan over the button and know it's there, then return to the top of the filter panel if they wish to clear all their selections.

Additionally, the title of the filter panel updates with the number of selected filters, but per #9392, these updates are not announced to AT users.

---

## ❌ 🎨 The visual appearance of filters with sub-items does not convey the existence of those sub-items.

<img width="383" alt="Screenshot 2023-06-22 at 12 28 26 PM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/25532785/9b9e96fb-97b2-4596-8a98-8260ba29988a">

<img width="385" alt="Screenshot 2023-06-22 at 12 28 14 PM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/25532785/5f7786d1-8349-4b38-aab8-31b967b18a43">

Under Technologies, there is a list of sub-items to filter on. This is correctly announced to AT users with the use of `aria-expanded`, but there's no visual indication that sub-items exist until and unless a user clicks on Artificial Intelligence. We need to provide some sort of visual affordance that there are sub-items in there.

Additionally, and on a personal level, I'm not sure how I feel about some items having checkboxes and some not. Was there some design-specific reason to have two separate looks for filter items?

**Comment by andy-blum** [comment]
@areagan1030 and/or @oliviaflory can you take a look at the issues marked with 🎨 ?

**Comment by oliviaflory** [comment]
@andy-blum 

> ❓Sets of checkboxes need to be reachable using arrow keys

This is definitely a ? for me as well because according to the Carbon [checkbox accessibility](https://carbondesignsystem.com/components/checkbox/accessibility#keyboard-interactions) guidance, the checkboxes are all individual tab stops. And as you mentioned above, when I suggested moving to arrow keys Michael G said that would not be following the expected keyboard behavior.

> ❌ 🎨 The placement/design of the "clear" button needs to be re-examined.
When items are selected and filters are applied, a new button pops into the DOM allowing users to clear their selections. The location of this button, however, places it in a location that's not intuitive to assistive technology users, as they'd have to search in reverse to find it. We have two primary options to fix this:
>
> 1. Move the clear-selections button to the bottom of the filter panel. This would make it easier for AT users to discover the button after making their selections
> 2. Keep the clear-selections button in the DOM but disable it when no filters are applied. In this way, AT users can scan over the button and know it's there, then return to the top of the filter panel if they wish to clear all their selections.

>Additionally, the title of the filter panel updates with the number of selected filters, but per https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/9392, these updates are not announced to AT users.

Yes, here are my thoughts on the options:
1. I like that the users could discover the button immediately, and after they apply filters and view results, they could quickly get back to the filter panel and clear all filters
2. I like that the clear button appears at the bottom and allows the user to clear after applying any filters. This might actually be better for clearing results as well because the `shift`+`tab` would likely bring the user back to the bottom of the filter panel?

> ❌ 🎨 The visual appearance of filters with sub-items does not convey the existence of those sub-items.
Under Technologies, there is a list of sub-items to filter on. This is correctly announced to AT users with the use of aria-expanded, but there's no visual indication that sub-items exist until and unless a user clicks on Artificial Intelligence. We need to provide some sort of visual affordance that there are sub-items in there.

We could provide a caret to indicate the additional items. I've seen in ibm.com/products that there are occasionally more than just two levels (we've seen 5). So how we indicate the levels is tricky. AND how do we make the level selectable and also a dropdown (this is not possible with accordion right now)

> Additionally, and on a personal level, I'm not sure how I feel about some items having checkboxes and some not. Was there some design-specific reason to have two separate looks for filter items?

In #10132 I was considering if we should change the filter panel to a tree component because this allows for multiple levels _with_ visual indicators, has less tab stops for the keyboard user, etc. But...it's challenging because you normally wouldn't mix a multi select (checkboxes) and single select tree together. 

IBM.com/products needed a filter panel that mixed single select and multi select options, so that is how we ended up with a mix of visual styles. I have a ton of explorations on this from two years ago, I also tried out radio buttons which were shot down. 

I did some additional research and found Amazon does do something similar to what we are currently doing (no visual indicator for sub levels), but that doesn't mean it's correct. 

<img width="878" alt="Screenshot 2023-06-23 at 3 55 49 PM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/45692622/15719ea4-8bc6-45fb-b153-8899770d2969">
<img width="876" alt="Screenshot 2023-06-23 at 3 55 58 PM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/45692622/e6cf8133-e4bd-48bf-a5b2-d8086f1ab34b">
<img width="871" alt="Screenshot 2023-06-23 at 3 56 06 PM" src="https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/45692622/a6336711-e54d-4193-acac-9edc28d1568a">

**Comment by andy-blum** [comment]
> it's challenging because you normally wouldn't mix a multi select (checkboxes) and single select tree together. ... I also tried out radio buttons which were shot down.

I'd be curious to hear the reasoning as radio buttons are the _perfect_ solution for this. What if we used radio buttons but styled them to _look_ like checkboxes? Then only one could ever be selected and they would match the visuals of the multi-select

**Comment by mbgower** [comment]
I'm having challenges digesting this thread, so let me play this back to you. I hear 3 problems:
1. You want the ability to select multiple product types, but a screenreader user has complained they want to operate between the  checkboxes for those products with arrow keys
2. You have put the clear filter button at the top, but a screen reader user has complained that is not a great experience because they have to move up the page to activate
3. You want to mix a radio button group and a multi-select group in the same filter space

Are those accurate?
Assuming they are:
1) one does not use arrow keys to navigate between checkboxes unless they are in some kind of multi-select dropdown. I would not describe what I'm seeing here as a dropdown. Ergo, it's a set of checkboxes and Tab is used to navigate between them. My surprise at any screen reader requesting this suggests that there is something odd going on under the covers that is making them misunderstand what I'm seeing visually. A call with everyone might help
2) I question the interaction, as described by the screen reader user. Here is what I assume happens here... 
 a) I set a filter
 b) I review the results
 c) if I don't like the filter, I clear it and start again OR I modify.
That flow seems reasonable to me, and the clear button at the top of the filter seems fine. Shoving it at the bottom of the set of filters seems even worse, as I'd have to navigate to the bottom of them to clear, then navigate back to the top to set the filter.
3) We'd better talk about the mixed filter space

**Comment by andy-blum** [comment]
@mbgower 

> You want the ability to select multiple product types, but a screenreader user has complained they want to operate between the checkboxes for those products with arrow keys

That's correct, agreed that a call with all parties might be best.

> You have put the clear filter button at the top, but a screen reader user has complained that is not a great experience because they have to move up the page to activate

It's not that the button is at the top, it's that it's not in the DOM at all until a selection is made, then the button is added in an area the AT user has already passed through, so they might not know it exists at all.

**Comment by andy-blum** [comment]
Agreed, and that's in the main issue description:

> 2. Keep the clear-selections button in the DOM but disable it when no filters are applied. In this way, AT users can scan over the button and know it's there, then return to the top of the filter panel if they wish to clear all their selections.

I probably should have been clearer about leaving it in the DOM _where it is_.

**Comment by oliviaflory** [comment]
Review 6/27

Entire filter panel is list item
- Categories are list items
   - Product types, Technologies is also a list item
### Checkboxes reachable by arrow keys
Currently:
- All items are a tab stop
- JAWS cannot traverse through checkboxes with arrow keys
- Ram wanted BOTH tab and arrow keys
- Only time MG has seen checkboxes NOT a tab stop, is in a multi-select dropdown

### Placement of clear 
- Placement
   - Could place on top AND bottom
   - "Clear filter" might be better context
- Clear button disabled on load
   - Will not be in tab order on load if disabled
   - Should we announce the button is enabled once a selection has been made?
- UX
   - When you click "Clear" the button will be disabled, where does the focus go? 
      -  Cannot be placed on a disabled button
   - Should Filter be a region?
      - Could be an elegant way to allow user to switch between filters and results
   
### Where to place number of results
- Should "Filter" change when a filter is applied
   - Filters
   - Filter results
   - Filtered by
   - Would be good to have results 

### Single and multi-select
- Radio buttons for single select
   - We didn't go with that direction because it was busy and would force an "All" to be available
   - Don't want radio buttons because your navigation is locked to your selection and would update content constantly and would be bad for the user.
- Possibly switch to Tree view
   - Can we mix multi-select and single select trees?
   - Could use multiple trees:
      - Single select
      - Single select
      - Multi select
      - Single select

### Results
- Consider having category sections [Technologies] etc to include the number of results
   - Technologies (150)
      - When a selection is made within Technologies it updates to Technologies (100)

### Colors
Olivia & Amanda to sync
- Single item selection state
   - Move away from blue (go for gray)
- Filter panel background
   - Could be white instead of g10
   - With a divider between results?
